### ISSUE TITLE
explore workbox support

### ISSUE THREAD

**Issue by hanford** [hanford/next-offline#3]
Both #2 #1 are experiencing some bugs that appear to be related Next.js + sw-precache, I've decided to explore alternative approaches and it appears there are many.

I'm thinking about supporting both [sw-precache](https://github.com/GoogleChromeLabs/sw-precache) and [workbox](workboxjs.org) ..

In the `workbox` branch I almost have workbox running 100%, I'll need to clean up this [script](https://github.com/hanford/next-offline/blob/workbox/precache.js) but the tl;dr is that we need to run a script after `next-build` that will go and grab all of the built bundles/chunks/pages and write them to a service worker that we generate .. this is tricky because next is also fingerprinting assets, so we need to adhere to the custom path. (that was the hard part, and it's mostly done minus some cleanup)

At my work, I have an [environment](https://eazecom-wjzxrxbsmo.now.sh) where I was able to get this implementation working 100% (opening the network work tab on first visit yeilds all assets next generates getting cached) .. I just need to generalize it a little bit before I cut a new release of `next-offline` 

I have the the sw generation working in this [repo](https://github.com/hanford/with-workbox) .. the only gotcha is running `node_modules/next-offline/precache.js` after next build. 

cc @OrangeXC @gerhardsletten

**Comment by hanford** [comment]
I've fixed the precache script so now it should work in all next.js applications ...

I also have two environments running [one with precache](https://with-sw-precache-rkrcakdmuq.now.sh/) and [one with workbox](https://with-workbox-rfrtrxevxf.now.sh/)

**Comment by gerhardsletten** [comment]
@hanford Hi, have been looking into this lately as well, and I have resolved it my own nextjs site. Original thought to cache (with precache) assets from `next build` but there where 2 drawbacks:

* Inspecting next's Head and NextScripts component reveal that reveals that js-files are served in a different, so just replacing `.next` with `_next` will not work.
* Also I had inline images I would like to cache as well and paths for those are served via a small graphql-service.

So landed on using continue using of toolbox-cache, but get it so cache my initial resources after the service-worker takes control.

First I created an [extra service-worker file](https://github.com/gerhardsletten/metabits-frontend/static/sw-inital-request.js) that could listened to calls from client and would take a list of urls and interact with the toolbox-cache directly:

```js
self.addEventListener('message', function (event) {
  if (event.data.assets) {
    for (var url of event.data.assets) {
      self.toolbox.cache(url)
    }
  }
})
self.addEventListener('install', function (event) {
  return self.skipWaiting()
})
self.addEventListener('activate', function (event) {
  event.waitUntil(clients.claim())
})
```
Its loaded with the `importScripts` option for `SWPrecacheWebpackPlugin` [soruce](https://github.com/gerhardsletten/metabits-frontend/blob/master/next.config.js#L33)

Then in my _document file I would send that list of urls to the method in the worker [source](https://github.com/gerhardsletten/metabits-frontend/blob/master/pages/_document.js#L37)

```js
{prod && (
  <script dangerouslySetInnerHTML={{__html: `
    if ('serviceWorker' in navigator) {
      navigator.serviceWorker
      .register('/service-worker.js')
      .then(registration => {
        return new Promise(resolve => {
          if (navigator.serviceWorker.controller) return resolve()
          navigator.serviceWorker.addEventListener('controllerchange', e => resolve())
        })
      })
      .then(() => {
        navigator.serviceWorker.controller.postMessage({
          assets: window.INITAL_CACHE
        })
      })
      .catch(err => {
        console.warn('service worker registration failed', err.message)
      })
    }
  `}} />
)}
```

That list of resources from initial request came from a `NextAssets` component supporting a render-callback (mainly just took it from Next's Head-component) [source](https://github.com/gerhardsletten/metabits-frontend/blob/master/elements/NextAssets.js)

```js
<NextAssets>
  {(assets) => {
    return (
      <script dangerouslySetInnerHTML={{__html: `
        window.INITAL_CACHE = ${JSON.stringify([...fixAssets(assets), pathname])};
      `}} />
    )
  }}
</NextAssets>
```

*Else* 

Original I had images laying and loaded directly from a S3 bucket, but had and issue sw-precache/toolbox Request not supporting cors-calls so I had to add a nginx reversed proxy to my bucket to have it on the same domain, you can copy the generated sw service-worker file and add a [mode: cors to its request](https://developer.mozilla.org/en-US/docs/Web/API/Request/mode), but then you would need to manage this file yourself..

**Comment by hanford** [comment]
@gerhardsletten FWIW my workbox stuff from ^ caches everything (api calls, assets, pages, etc) .. even requests to third parties (like google!) .. the script I wrote above works, but it needs to execute after build .. and it will correctly inline all `js` assets into the precache sw. An example on a relatively complex website can be seen [here](https://sw-test.now.sh).

I think this should solve your problem. I'll probably wrap this up and rewrite the README / cut a release tonight. I'll also update my personal website with the new service worker code

workbox solves most of the issues that sw-precache wasn't ..

**Comment by gerhardsletten** [comment]
Looks good, but if I understands correctly everything is controlled by this script [sw-test.now.sh/service-worker.js](https://sw-test.now.sh/service-worker.js)?

One thing I wonder about is the path's NextJS serve for its app, pages and chunks  if it would be possible for NextJS to expose them in some way.

And yes, Workbox as a descender of sw-precache is the way to go

**Comment by hanford** [comment]
@gerhardsletten shipped `1.0.0` I've updated my personal website with the newest implementation and you can see pre-caching app, pages and chunks by default on first visit. (Might need to unregister my service worker and hard refresh)
